### DISCUSSION TITLE
Some resource of driver from linux (Kernel 5.4.0)

### DISCUSSION THREAD

**Discussion by quynkk1** [SkyrilHD/VoodooPS2-ALPS#4] [General]
I found some resource of linux kernel from Linux 5.4.0 (Support V1 - V8) and I want you to review it.
I changed some part in your code, but nothing happened (No movement of cursor, Trackpad setting show not detected) and when I reboot, it got panicked.
My laptop is Dell Latitude E5570 with ALPS V8
(Sorry for my bad EN)
https://gitlab.sdu.dk/sdurobotics/linux-kernels/kernel/-/blob/focal/Ubuntu-5.4.0-31.35/drivers/input/mouse/alps.c

**Comment by SkyrilHD** [comment]
The ALPS code is already from the Linux kernel. I added partial support for V8 as well, so you could try without changing the code. Does it work with the pre-VoodooInput kext ([v1.0.0](https://github.com/SkyrilHD/VoodooPS2-ALPS/releases/tag/1.0.0))? If so, I have an idea what causes the issue.

Also what did you change in the code?

**Reply by quynkk1** [reply]
I put your kext, but all keyboard, trackpad, trackstick stop working.
I added trackpad detect finger and i think i could move but no success. It got panic when I reboot.
The kext I uploaded can use both trackpad and trackstick but no setting in SysPref, just support tap with 1-2 fingers

[VoodooPS2Controller.zip](https://github.com/SkyrilHD/VoodooPS2-ALPS/files/7713086/VoodooPS2Controller.zip)

**Reply by SkyrilHD** [reply]
The kext you uploaded is from acidanthera. The reason why you got no settings in System Preferences is because you use VoodooPS2Mouse instead of VoodooPS2Trackpad. You can check it by disabling VoodooPS2Mouse in your config.

Keyboard does not work with 1.0.0 or 1.0.4?

**Reply by quynkk1** [reply]
Both 1.0.0 and 1.0.4 keyboard stop working and trackpad, too

**Reply by SkyrilHD** [reply]
Logs would be very useful. But your keyboard should work. Maybe I can see why it doesn't work in the logs

**Comment by quynkk1** [comment]
You can check my video and logs about testing your kext. I attached my log about system.log + boot.log

[log.zip](https://github.com/SkyrilHD/VoodooPS2-ALPS/files/7721177/log.zip)

https://user-images.githubusercontent.com/95995307/146227252-f41d7513-1c05-4107-9449-4271ef21c5c7.mp4

**Reply by SkyrilHD** [reply]
Checking your logs I found this. `org.rehabman.voodoo.driver.PS2Controller error - OSBundleCompatibleVersion 2.8.15 > CFBundleVersion 1.0.0 (must be <=).`

Can you try this kext: [VoodooPS2Controller-1.0.0-DEBUG.zip](https://github.com/SkyrilHD/VoodooPS2-ALPS/files/7721680/VoodooPS2Controller-1.0.0-DEBUG.zip)
 
This kext is working for me. 
![image](https://user-images.githubusercontent.com/28839925/146239428-0bb92cd6-9ea4-4e81-8df8-36f62ef64e3a.png)

**Reply by quynkk1** [reply]
Sorry for late reply.
I have tested your VoodooPS2 Debug kext, but it's still nothing about movement, keyboard not responding...
I will generate log file later

**Comment by quynkk1** [comment]
This is my log which I dumped from Hackintool.
[Log.zip](https://github.com/SkyrilHD/VoodooPS2-ALPS/files/7745929/Log.zip)
https://user-images.githubusercontent.com/95995307/146782357-40776040-83f0-40d2-ab28-183584a598bd.mp4

**Reply by SkyrilHD** [reply]
I looked at your logs and did not find anything useful. There is nothing about initializing VoodooPS2. You may want to increase the kernel's buffer size by adding this to the boot-args: `msgbuf=1048576`.

I enabled logging and processing for trackstick and enabled more logging for Trackpad: [VoodooPS2Controller-1.0.0-DEBUG.zip](https://github.com/SkyrilHD/VoodooPS2-ALPS/files/7747032/VoodooPS2Controller-1.0.0-DEBUG.zip)

**Reply by quynkk1** [reply]
Trackpad still cannot move the cursor, but I can click Left/Right
Trackpoint move the cursor but the cursor went to bottom-right of screen, Left/Middle/Right Click still nothing
Keyboard is working well.
Please check the log I sent.

[Log E5570 ALPS v8.zip](https://github.com/SkyrilHD/VoodooPS2-ALPS/files/7763714/Log.E5570.ALPS.v8.zip)

**Reply by SkyrilHD** [reply]
So, looking at your log, I can see you get inputs. However, since you use V8+ the processing is not correct, which (I think) has been corrected in the latest release of the kext. I edited the track point calculation and passed touchpad input directly to the system, as it only passed dx:0 and dy:0. Theoretically, you should get an input with the touchpad.

`[   80.685354]: ALPS: SS4_PACKET_ID_ONE`
`[   80.685361]: ALPS: SS4_V2: x:1518 y:3284 pressure:22, fingers:1 buttons:0New buttons value after check for middle click: 0`
`[   80.685368]: VoodooPS2::Mode: 1`
`[   80.685369]: PS2: Cancel DX: 0 Cancel DY: 0PS2:Checking Fingersps2: fingers=1, dx=0, dy=0 (0,0) z=22 mode=(1,1,1) buttons=0 wasdouble=0 wastriple=0`

You can see that x:1518 y:3284 is being passed but only returns dx=0, dy=0 to the system. Also, the coordinates are too high, I guess it is due to the different calculation done between V8 and V8+.

EDIT: Forgot to upload the kext: [VoodooPS2Controller-1.0.0-DEBUG.zip](https://github.com/SkyrilHD/VoodooPS2-ALPS/files/7765137/VoodooPS2Controller-1.0.0-DEBUG.zip)

**Comment by SkyrilHD** [comment]
@quynkk1 Can you try the newest master kext? It has V8 support.

**Reply by mrockwood** [reply]
Not OP but I just gave it a try on my Latitude 7370, which has a V8 trackpad. The kext works great until the laptop goes to sleep. After waking the trackpad is frozen, but the keyboard continues to work.

**Reply by SkyrilHD** [reply]
Can you try this? It resets the touchpad before initializing.
[VoodooPS2Controller-1.0.5-DEBUG.zip](https://github.com/SkyrilHD/VoodooPS2-ALPS/files/7797927/VoodooPS2Controller-1.0.5-DEBUG.zip)

Does it happen every time it wakes up or now and then?

**Reply by mrockwood** [reply]
The attached kext still freezes. Of the 10-12 attempts I have made on my machine, it has happened every time.

**Reply by SkyrilHD** [reply]
Can you make the WakeDelay larger (instead 1000 e.g. do 5000)? (It is in VoodooPS2Controller.kext>Contents>PlugIns>VoodooPS2Trackpad.kext>Contents>Info.plist>IOKitPersonalities>ALPS GlidePoint)

![image](https://user-images.githubusercontent.com/28839925/147862205-5d5a19e8-95a9-428a-adca-7bd1e2c53655.png)

**Reply by mrockwood** [reply]
Setting WakeDelay to 5000 doesn't change the trackpad and appears to freeze the keyboard after waking as well.

**Reply by SkyrilHD** [reply]
Can you test the latest master version? I had two people confirm the issue is now fixed.

https://github.com/SkyrilHD/VoodooPS2-ALPS/suites/4817644541/artifacts/135763243

**Reply by mrockwood** [reply]
The debug kext seems to be working! Thanks for working on this.

**Reply by quynkk1** [reply]
Sorry for my late reply because of my examination in class.
I have tested, both trackpad and trackpoint working well.
Thanks for working on this.

**Comment by quynkk1** [comment]
After used your new kext for 30mins, I found this bug.

https://user-images.githubusercontent.com/95995307/148097648-9a693428-b8d7-45db-8d77-fe079d409048.mp4

**Reply by SkyrilHD** [reply]
Gather logs and send them

**Reply by quynkk1** [reply]
[Archive.zip](https://github.com/SkyrilHD/VoodooPS2-ALPS/files/7811974/Archive.zip)
Log attached.

**Reply by SkyrilHD** [reply]
Looks like it is stuck to these values:

![image](https://user-images.githubusercontent.com/28839925/148226278-781b3d07-2873-4515-a473-9b4f416e04c1.png)

x:0 y:1536 and 0 fingers

What does it look like in FingerMgmt?

**Reply by quynkk1** [reply]
Still that problem.
After using kext (DEBUG for detecting error) for 3-5 mins, cursor randomly jump and automatic set y=0 at finger placed on trackpad, so you can move 2 side left and right (or horizontal) but cannot move up/down (or vertical).
With 4-finger, I can swipe for Launchpad or Show desktop.
Any idea or solution for that?
Note: I'm using macOS Monterey 12.1 with v8 ALPS

**Reply by SkyrilHD** [reply]
Logs would be helpful and what was the last thing you did before it happened?  I think after doing something the touchpad gets stuck in a mode that disables vertical movements.

Test it like this:
Use 1 finger for a few minutes
No issues? Then, try 2 fingers for a few minutes
No issues? Then, try 3 fingers for a few minutes
Etc.

It is easier if we know where the problem is.

**Reply by quynkk1** [reply]
[Log.zip](https://github.com/SkyrilHD/VoodooPS2-ALPS/files/7832759/Log.zip)

1-4 finger for a few minutes after boot is fine, no problem
After used for 5-10 minutes, I moved with 1 finger, cursor jump and cannot move around but I can move horizontal as a line
Trackpoint at that time still good, can move around without problem.

**Reply by SkyrilHD** [reply]
What does it look like on FingerMgmt?

**Reply by quynkk1** [reply]
Check my log and video here.
https://drive.google.com/file/d/1RvUzQ0qr2sF8QHp02LfFXmnWlBtWUz66/view?usp=sharing

**Reply by SkyrilHD** [reply]
Thanks. Now I roughly understand what happens. Will look into it.

Btw, did you try with the RELEASE version?

Also, can you try remapping the Print screen button with this SSDT: https://github.com/acidanthera/VoodooPS2/blob/master/Docs/ACPI/SSDT-PrtSc-Remap.dsl (You need to compile it first), so when you press the Print screen button, the trackpad gets disabled. Press the Print screen button again to enable the trackpad and see if that changes anything. If it does not change anything, try resetting the trackpad by pressing Ctrl+Alt+PrntScr.

**Reply by quynkk1** [reply]
Same problem with Release version 1.0.5.
I added SSDT-PrctSc-Remap.aml into Opencore, after reboot 3 minutes trackpad went into problem as I told you.
First, I pressed PrtScr twice: nothing changed, cursor moved as a line.
Second, I pressed combination of Ctrl + Alt + PrtScr, both Trackpad and Trackpoint stop working until you restart your hack, it worked back.

**Reply by SkyrilHD** [reply]
Can you try this kext?

[VoodooPS2Controller-1.0.6-DEBUG.zip](https://github.com/SkyrilHD/VoodooPS2-ALPS/files/7857417/VoodooPS2Controller-1.0.6-DEBUG.zip)

**Reply by quynkk1** [reply]
I have tested that kext downloaded from [OSXLatitude](https://osxlatitude.com/forums/topic/17005-alps-touchpad-driver-with-voodooinput/?do=findComment&comment=113399) and it's worked well. Here is the log.
[log.txt](https://github.com/SkyrilHD/VoodooPS2-ALPS/files/7864804/log.txt)

**Reply by quynkk1** [reply]
After using this kext for 6-12 hours (because I re-installed macOS and xCode), cursor jump and got panic. Image:
![image](https://user-images.githubusercontent.com/95995307/149461381-da687923-d7e4-48e8-b425-3baa6693ca4c.png)

panic(cpu 2 caller 0xffffff8013bd78d3): Kernel trap at 0xffffff801933aa26, type 14=page fault, registers:
CR0: 0x000000008001003b, CR2: 0xffffff9508c781fc, CR3: 0x00000000237cc000, CR4: 0x00000000003626e0
RAX: 0xffffff9508c767b8, RBX: 0x00000000ffffffff, RCX: 0x000000000000056d, RDX: 0x0000000000000266
RSP: 0xffffffd050d8bdb0, RBP: 0xffffffd050d8bdb0, RSI: 0x000000000000056d, RDI: 0xffffff9508c76800
R8:  0x0000000000000000, R9:  0xfffffffffffffff7, R10: 0x0000000000000000, R11: 0xffffff8b6f8bd540
R12: 0xffffff9508c76800, R13: 0x0000000000000000, R14: 0x0000000000000000, R15: 0x0000000000000001
RFL: 0x0000000000010202, RIP: 0xffffff801933aa26, CS:  0x0000000000000008, SS:  0x0000000000000000
Fault CR2: 0xffffff9508c781fc, Error code: 0x0000000000000000, Fault CPU: 0x2, PL: 0, VF: 10

Panicked task 0xffffff903bb48670: 171 threads: pid 0: kernel_task
Backtrace (CPU 2), panicked thread: 0xffffff8b6f8bd540, Frame : Return Address
0xffffffd050d8b7e0 : 0xffffff8013a87ddd 
0xffffffd050d8b830 : 0xffffff8013be7d15 
0xffffffd050d8b870 : 0xffffff8013bd74e3 
0xffffffd050d8b8c0 : 0xffffff8013a27a60 
0xffffffd050d8b8e0 : 0xffffff8013a881ad 
0xffffffd050d8ba00 : 0xffffff8013a87966 
0xffffffd050d8ba60 : 0xffffff80143160b9 
0xffffffd050d8bad0 : 0xffffff8013bd78d3 
0xffffffd050d8bc50 : 0xffffff8013bd75b8 
0xffffffd050d8bca0 : 0xffffff8013a27a60 
0xffffffd050d8bcc0 : 0xffffff801933aa26 
0xffffffd050d8bdb0 : 0xffffff8019339863 
0xffffffd050d8be20 : 0xffffff801933879e 
0xffffffd050d8bea0 : 0xffffff801933c30c 
0xffffffd050d8bed0 : 0xffffff8014258d13 
0xffffffd050d8bf20 : 0xffffff801425756e 
0xffffffd050d8bf60 : 0xffffff8014256b97 
0xffffffd050d8bfa0 : 0xffffff8013a2718e 
      Kernel Extensions in backtrace:
         com.skyrilhd.PS2Trackpad(1.0.6)[D646DE0C-2093-3086-8207-D2CC1E162E68]@0xffffff8019334000->0xffffff8019344fff
            dependency: as.acidanthera.voodoo.driver.PS2Controller(1.0.6)[AC164ECA-3F5C-3209-9FDC-660CD5CCCD91]@0xffffff80192f5000->0xffffff801930cfff
            dependency: com.apple.iokit.IOHIDFamily(2.0.0)[4773BB64-73AE-3283-8DE4-D10459C90231]@0xffffff801627d000->0xffffff80162dafff

Process name corresponding to current thread (0xffffff8b6f8bd540): kernel_task
Boot args: forceRenderStandby=0 rtsx_sleep_wake_delay_ms=1000 -noht40

Mac OS version:
21C52

Kernel version:
Darwin Kernel Version 21.2.0: Sun Nov 28 20:28:54 PST 2021; root:xnu-8019.61.5~1/RELEASE_X86_64
Kernel UUID: E32C293C-7D8B-327C-A335-6AAD250A4881
KernelCache slide: 0x0000000013800000
KernelCache base:  0xffffff8013a00000
Kernel slide:      0x0000000013810000
Kernel text base:  0xffffff8013a10000
__HIB  text base: 0xffffff8013900000
System model name: MacBookPro13,2 (Mac-66E35819EE2D0D05)
System shutdown begun: NO
Panic diags file available: YES (0x0)
Hibernation exit count: 0

System uptime in nanoseconds: 43691488736858
Last Sleep:           absolute           base_tsc          base_nano
  Uptime  : 0x000027bcb7aa7b06
  Sleep   : 0x00000046f3a05765 0x00000008e3f221ce 0x0000000000000000
  Wake    : 0x0000004703b50e02 0x00000000a551e3ee 0x00000046fd47f20e
Zone info:
Foreign   : 0xffffff8023fce000 - 0xffffff8023fdb000
Native    : 0xffffff803bcde000 - 0xffffffa03bcde000
Readonly  : 0xffffff85089aa000 - 0xffffff86a233e000
Metadata  : 0xffffffe921cca000 - 0xffffffe941e48000
Bitmaps   : 0xffffffe941e48000 - 0xffffffe947e48000

**Comment by quynkk1** [comment]
I have tested this kext on Dell Latitude E5440 from my friend, it worked well until restart computer and stop working both trackpad and keyboard.
OS I tested: macOS Catalina, 4GB DDR3L, 120G SSD, v4 ALPS(?, I think that)

**Reply by SkyrilHD** [reply]
You can check the version by checking the IOReg:

![image](https://user-images.githubusercontent.com/28839925/149742040-d9d0f7f5-36ed-4115-b16d-66b1ca180598.png)

Does he have SMCDellSensors? It causes issues with ALPS touchpads.

If not, logs would be appreciated.

**Reply by quynkk1** [reply]
I have removed SMCDellSensor and it worked like a charm. Thanks

**Reply by SkyrilHD** [reply]
Which ALPS version is it btw?

**Reply by quynkk1** [reply]
I think that it was v3 or v4...

**Reply by SkyrilHD** [reply]
If the E5440 has the same touchpad as the E7240, which is E7=0x73 0x03 0x0a, EC=0x88 0x08 0x22, then it is a Rushmore V3. But I am not sure, it would be better if you check the version as the instructions I explained here: https://github.com/SkyrilHD/VoodooPS2-ALPS/discussions/4#discussioncomment-1981787 The property should show 0x310 to confirm it is a Rushmore touchpad.

Does everything work?

**Reply by quynkk1** [reply]
Everything is working with your kext on my E5440, but I have a question: Does NumLock support in this build?

**Reply by SkyrilHD** [reply]
It should. However, one needs to add an SSDT in order to make NumLock to work.

https://github.com/acidanthera/VoodooPS2/blob/master/Docs/ACPI/SSDT-NumLockSupport.dsl

**Reply by quynkk1** [reply]
I have attached SSDT-NumLockSupport.aml but after restart (I have tried Reset NVRAM), NumLock still no work.
Pressed Numlock and press number 4 (Left Button) but it shown number 4 instead of moving cursor to left

**Reply by SkyrilHD** [reply]
Check in IOReg if NumLockSupport is True:

![image](https://user-images.githubusercontent.com/28839925/150943826-c859a246-238a-49e0-abf3-b2a1820bcc38.png)

Maybe try holding down the Fn key.

I have a feeling these buttons don't work in macOS.

**Reply by quynkk1** [reply]
![image](https://user-images.githubusercontent.com/95995307/150948529-b29310df-8910-464e-b470-717026d5b377.png)

NumLockSupport in IOReg is True
Holding down Fn key with NumLock and pressed direction still print number instead of moving text cursor.
